### ISSUE TITLE
More documented/commented source code.

### ISSUE THREAD

**Issue by KoxaKoxama** [godotengine/godot-proposals#3484]
### Describe the project you are working on

Upon reading Godot source code to find and understand how certain features work, sometimes it's hard to get what each .cpp files is meant for and how they are interacting with other parts of the engine. 
There is simply little to no explanation given in documentation.

Although it's a good practice to write c++ code in understandable manner so that there is no need for additional comments, for large open source project like this, it's beneficial to have as much explanation baked into comments as it could possibly be, even if things looks obvious for main developers.

### Describe the problem or limitation you are having in your project

Let's face it, literally any game made with Godot can be cracked in two clicks and unpacked back to it's source code, even project.binary can be reversed back to project.godot and opened directly in Godot Editor. There are tools for that, and there will be more. 
This is unacceptable for any commercial product made with Godot.
Companies will eventually try to use Godot for their games, and they will soon realize that at least some form of protection is needed.

Although script encryption (and in the future release of Godot 4.0 whole pck enctyption) is implemented, this is completely useless for actual protection, as literally anyone with mediocre knowledge of tools like x64dbg can trace encryption key in executable and use it to unpack any game back to working source code and open and play it directly in Godot Editor.

Therefore the best protection developers can use, is abbility to modify engine source code, espesialy files related to pck such as pck_packer.cpp, file_access_pack.cpp, and others, and this is why they must have well commented code for ease of use. 
Source code should work as a guide into how the engine works, with all warnings and suggestions clearly written in comments.

**Comment by pycbouh** [comment]
While some parts of the codebase could definitely benefit from more explanations, this proposal is too vague to do anything about it. Since your problem is about being able to easily customize the part of the engine that handles the encryption of the assets, I suggest you make a proper proposal about how the engine must be changed to allow that, even if by "allow that" it meant "making it easier to change in forks".

As for the problem itself, do keep in mind that any project that has to run on the end-user machine can be cracked quite easily. And even DRM systems have proven useless and only harming user experience and sales. So take caution with where your journey in this regards leads you.

**Comment by CedNaru** [comment]
Documenting a good part of the source code is infeasible by now. But one thing I would like to see (I'm working on Modules mostly) is better documentations for the macros used in C++.
There is already a page in the documentation about them, but the list is far from exhaustive and the current master branch introduced a fair amount of new macros. 

Many times when I am navigating the source code of the engine, I meet some new macro I didn't know about before and which usage is unclear. Expanding them to see the source code doesn't help soemtimes because they are a cascade of other macros/templates/variadic arguments/black magic.

Or more simply, sometimes I write my own helper macros/functions for my module just to discover that they was already a macro somewhere in the engine doing exactly that.

Bottom Line: Improving macro documentation woud make writing code and understanding it faster.

**Comment by Calinou** [comment]
Documentation requests should be made on the [godot-docs](https://github.com/godotengine/godot-docs) repository, not godot-proposals. That said, remember to keep your issues focused to a specific area. "Make the source code more documented" isn't an actionable proposal in itself, as it's too general.

Also, remember that requesting someone to document a file they're not familiar with is a bad idea. Superficial comments that repeat the source code aren't helping anyone.
